hydromt/components/mesh.py

```diff
@@ -1,40 +1,384 @@
+    def mesh_gdf(self) -> Dict[str, gpd.GeoDataFrame]:
+        """Returns dict of geometry of grids in mesh as a gpd.GeoDataFrame."""
+        mesh_gdf = dict()
+        if self.data is not None:
+            for k, grid in self.mesh_grids.items():
+                if grid.topology_dimension == 1:
+                    dim = grid.edge_dimension
+                elif grid.topology_dimension == 2:
+                    dim = grid.face_dimension
+                gdf = gpd.GeoDataFrame(
+                    index=grid.to_dataset()[dim].values.astype(str),
+                    geometry=grid.to_shapely(dim),
+                )
+                mesh_gdf[k] = gdf.set_crs(grid.crs)
+
+        return mesh_gdf
+
+    def get_mesh(
+        self, grid_name: str, include_data: bool = False
+    ) -> Union[xu.Ugrid1d, xu.Ugrid2d, xu.UgridDataArray, xu.UgridDataset]:
+        """
+        Return a specific grid topology from mesh based on grid_name.
+
+        If include_data is True, the data variables for that specific
+        grid are also included.
+
+        Parameters
+        ----------
+        grid_name : str
+            Name of the grid to return.
+        include_data : bool, optional
+            If True, also include data variables, by default False.
+
+        Returns
+        -------
+        uds: Union[xu.Ugrid1d, xu.Ugrid2d, xu.UgridDataArray, xu.UgridDataset]
+            Grid topology with or without data variables.
+        """
+        if len(self.data) < 1:
+            raise ValueError("Mesh is not set, please use set_mesh first.")
+        if grid_name not in self.mesh_names:
+            raise ValueError(f"Grid {grid_name} not found in mesh.")
+        if include_data:
+            # Look for data_vars that are defined on grid_name
+            variables = []
+            for var in self.data.data_vars:
+                if hasattr(self.data[var], "ugrid"):
+                    if self.data[var].ugrid.grid.name != grid_name:
+                        variables.append(var)
+                # additionnal topology properties
+                elif not var.startswith(grid_name):
+                    variables.append(var)
+                # else is global property (not grid specific)
+            if variables and len(variables) < len(self.data.data_vars):
+                uds = self.data.drop_vars(variables)
+                # Drop coords as well
+                drop_coords = [c for c in uds.coords if not c.startswith(grid_name)]
+                uds = uds.drop_vars(drop_coords)
+            elif variables and len(variables) == len(self.data.data_vars):
+                grid = self.mesh_grids[grid_name]
+                uds = xu.UgridDataset(grid.to_dataset(optional_attributes=True))
+                uds.ugrid.grid.set_crs(grid.crs)
+            else:
+                uds = self.data.copy()
+
+            return uds
+
+        else:
+            return self.mesh_grids[grid_name]
+
+    @hydromt_step
+    def add_data_from_rasterdataset(
```

Comment:
I still miss in the name that this method works for 2D grids and not 1D and 3D. I previously suggested ``add_2ddata_from_rasterdataset``. I'm never good with names but I still think we should highlight this fact somehow in the name.

Reply:
I renamed it to `add_2d_data_from_rasterdataset`